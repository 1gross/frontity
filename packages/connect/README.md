## Description
The state manager of Frontity. Similar to MobX but much more lightweight because it's based on ES6 proxies.

You can find more information about it at [Frontity docs](https://docs.frontity.org/javascript-and-and-react/react-basic#state-manager-frontity-connect).
