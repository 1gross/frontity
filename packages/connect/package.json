{
  "name": "@frontity/connect",
  "version": "0.1.0",
  "description": "A state manager that's easy to use and hackable at the same time.",
  "keywords": [
    "frontity",
    "connect"
  ],
  "homepage": "https://frontity.org",
  "license": "Apache-2.0",
  "repository": {
    "type": "git",
    "url": "https://github.com/frontity/frontity.git"
  },
  "bugs": {
    "url": "https://community.frontity.org"
  },
  "scripts": {
    "test:ci": "jest",
    "test": "jest --watch",
    "test:inspect": "node --inspect ./node_modules/.bin/jest --watch --no-cache --runInBand"
  },
  "dependencies": {
    "recursive-iterator": "^3.3.0"
  },
  "devDependencies": {
    "@babel/core": "^7.4.4",
    "@babel/preset-env": "^7.4.4",
    "@types/jest": "^24.0.11",
    "babel-jest": "^24.8.0",
    "jest": "^24.7.1",
    "ts-jest": "^24.0.2",
    "typescript": "^3.4.2"
  }
}
