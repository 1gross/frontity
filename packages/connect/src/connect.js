import { createContext, useContext, createElement } from "react";
import { view } from "@risingstack/react-easy-state";

/**
 * A React context that stores the store of Frontity Connect.
 */
const context = createContext();

/**
 * The Provider of the store context, to be used in the root of the React
 * application.
 */
export const Provider = context.Provider;

/**
 * Connects a React component with the Frontity connect store.
 *
 * @param Component The React component to be connected
 * @returns The same Component, but with the store included in the props.
 */
export const connect = (Component) => {
  const isStatelessComp = !(
    Component.prototype && Component.prototype.isReactComponent
  );

  if (isStatelessComp) {
    return view((props) => {
      const store = useContext(context);
      return Component({ ...props, ...store });
    });
  } else {
    class ConnectedComponent extends Component {
      static contextType = context;

      render() {
        return createElement(Component, {
          ...this.props,
          ...this.context,
        });
      }
    }

    return view(ConnectedComponent);
  }
};
