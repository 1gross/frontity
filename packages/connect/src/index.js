export { observe, unobserve } from "./observer";
export { observable, isObservable, raw, clone } from "./observable";
export { createStore, getSnapshot } from "./create-store";
export { connect as default, Provider, useConnect } from "./connect";
export { batch } from "./scheduler";
