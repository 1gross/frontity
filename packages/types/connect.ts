import Package from "./package";
import { ResolveState, Omit } from "./utils";

// Resolves the actions for things like Action and Derived.
export type ResolveActions<Actions extends Package["state"]> = {
  [P in keyof Actions]: Actions[P] extends (
    ...a: any
  ) => (arg: infer Arg) => void // Turns "state => args => {}" into "args => {}"
    ? (arg: Arg) => void // Turns "state => {}" into "() => {}"
    : Actions[P] extends (state: Package["state"]) => any
    ? () => void
    : ResolveActions<Actions[P]>
};

export type Connect<S extends Package, Props extends object = {}> = Omit<
  S,
  "state" | "actions" | "name"
> & {
  state: ResolveState<S["state"]>;
  actions: ResolveActions<S["actions"]>;
} & Props;

export default Connect;
