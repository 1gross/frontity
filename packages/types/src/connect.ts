import Package from "./package";
import { ResolveState, Omit, ResolveActions } from "./utils";

export type Connect<S extends Package, Props extends object = {}> = Omit<
  S,
  "state" | "actions" | "name"
> & {
  state: ResolveState<S["state"]>;
  actions: ResolveActions<S["actions"]>;
} & Props;

export type ExternalProps<T extends Package> = Omit<
  T,
  "state" | "actions" | "libraries" | "roots" | "fills"
>;

export default Connect;
