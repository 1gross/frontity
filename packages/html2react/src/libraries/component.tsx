/* eslint-disable @typescript-eslint/no-use-before-define */
import React from "react";
import { connect } from "frontity";
import { Connect } from "frontity/types";
import Html2ReactType, {
  Component,
  HandleNodes,
  HandleNode,
  ApplyProcessors
} from "../../types";

const applyProcessors: ApplyProcessors = ({ node, root, processors }) => {
  for (const processor of processors) {
    const { test: tester, process } = processor;
    let isMatch = false;

    // Test processor.
    try {
      isMatch = tester(node);
    } catch (e) {
      // Do nothing.
    }
    if (!isMatch) continue;

    // Apply processor.
    try {
      const processed = process(node, { root });
      // Return true if node was removed.
      if (!processed) return true;
      // Merge the nodes if the processor has applied changes.
      if (node !== processed) Object.assign(node, processed);
    } catch (e) {
      console.error(e);
    }
  }

  // Return false if node was not removed.
  return false;
};

const handleNode: HandleNode = ({ node, payload, index }) => {
  // `applyProcessors` returns true if node was removed.
  if (applyProcessors({ node, ...payload })) return null;
  if (node.type === "comment") return null;
  if (node.type === "text") return node.content;
  if (node.type === "element")
    return (
      <node.component {...{ key: index, ...node.props }}>
        {node.children ? handleNodes({ nodes: node.children, payload }) : null}
      </node.component>
    );
};

const handleNodes: HandleNodes = ({ nodes, payload }) => {
  const handled = nodes.reduce((final: React.ReactNodeArray, node, index) => {
    const handledNode = handleNode({ node, index, payload });
    if (handledNode) final.push(handledNode);
    return final;
  }, []);

  if (handled.length > 1) return handled;
  if (handled.length === 1) return handled[0];
  return null;
};

export const Html2React: Component<
  Connect<Html2ReactType, { html: string }>
> = ({ html, libraries }) => {
  const { processors, parse } = libraries.html2react;
  const root = parse(html);

  libraries.html2react.processors = processors.sort(
    (a, b) => (a.priority || 10) - (b.priority || 10)
  );

  return handleNodes({
    nodes: root,
    payload: {
      root,
      processors
    }
  }) as React.ReactElement;
};

export default connect(Html2React);
