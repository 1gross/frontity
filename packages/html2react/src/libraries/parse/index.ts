import { css, decode } from "frontity";
import { parse as himalaya } from "himalaya";
import { Element, Node, Parse, Attributes, AdaptNode } from "../../../types";
import htmlAttributes from "./attributes/html.json";
import svgAttributes from "./attributes/svg.json";

// Map of lowercased HTML and SVG attributes to get their camelCase version.
const attributesMap: Attributes = htmlAttributes
  .concat(svgAttributes)
  .reduce((map, value) => {
    map[value.toLowerCase()] = value;
    return map;
  }, {});

// Adapts the Himalaya AST Specification v1 to our format.
const adaptNode: AdaptNode = (himalayaNode, parent) => {
  let node: Node;

  if (himalayaNode.type === "element") {
    node = {
      type: himalayaNode.type,
      component: himalayaNode.tagName,
      props: himalayaNode.attributes.reduce(
        (props: Element["props"], { key, value }) => {
          if (key === "class") {
            props.className = value;
          } else if (key === "style") {
            // Add inline styles to the component with `emotion`.
            props.css = css(value);
          } else if (key === "href") {
            props.href = decode(value);
          } else if (key === "for") {
            props.for = value;
          } else if (!/^on/.test(key)) {
            const camelCaseKey = attributesMap[key.toLowerCase()];
            // Map keys with no value to `true` booleans.
            props[camelCaseKey || key] = value === null ? true : value;
          }

          return props;
        },
        {}
      )
    };

    node.children = himalayaNode.children.reduce(
      (tree: Node[], child): Node[] => {
        const childAdapted = adaptNode(child, node as Element);
        if (childAdapted) tree.push(childAdapted);
        return tree;
      },
      []
    );
  }

  if (himalayaNode.type === "text") {
    const content = decode(himalayaNode.content);

    if (content.trim().length) {
      node = {
        type: himalayaNode.type,
        content: content
      };
    } else return null;
  }

  if (himalayaNode.type === "comment") {
    const content = decode(himalayaNode.content);

    if (content.trim().length) {
      node = {
        type: himalayaNode.type,
        content: content
      };
    } else return null;
  }

  if (parent) node.parent = parent;

  return node;
};

const parse: Parse = html =>
  himalaya(html).reduce((tree: Node[], element) => {
    const adapted = adaptNode(element);
    if (adapted) tree.push(adapted);
    return tree;
  }, []);

export default parse;
