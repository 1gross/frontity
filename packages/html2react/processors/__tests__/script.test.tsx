import React from "react";
import { render } from "@testing-library/react";
import { createStore } from "@frontity/connect";
import { Html2React } from "../../src/libraries/component";
import Html2ReactPackage from "../../src";

const store = createStore(Html2ReactPackage);

jest.mock("@frontity/components/script", () => "mocked-script");

describe("Script processor", () => {
  it("should process a script with src", () => {
    const { container } = render(
      <Html2React html={'<script src="/some-js-file.js" />'} {...store} />
    );
    expect(container).toMatchInlineSnapshot(`<diva />`);
  });
  it("should process a script with code", () => {
    const { container } = render(
      <Html2React html={'<script>const some = "code";</script>'} {...store} />
    );
    expect(container).toMatchInlineSnapshot(`<div />`);
  });
});
