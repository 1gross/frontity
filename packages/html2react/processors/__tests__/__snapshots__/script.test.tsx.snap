// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Script processor should not process scripts with a non valid type 1`] = `
<div>
  <script
    src="/"
    type="non-valid"
  />
</div>
`;

exports[`Script processor should process a script with code 1`] = `
<div>
  <mocked-script
    code="const some = \\"code\\";"
  />
</div>
`;

exports[`Script processor should process a script with src 1`] = `
<div>
  <mocked-script
    src="/some-js-file.js"
  />
</div>
`;
